### PR TITLE
bump `eth-sig-util` version to 5.0.2, stop inheriting from `eth-simple-keyring`

### PR DESCRIPTION
**Pull request by adonesky1**
- Bumps version of `eth-sig-util` (now `@metamask/eth-sig-util`) to v5.0.2.

- Removes inheritance of `eth-simple-keyring` as we want to reorganize the relationships of various keyrings. The new structure will be that all keyrings, once converted to typescript, implement [a base-keyring type](https://github.com/MetaMask/types/pull/5) which indicates the required method signatures to be a valid keyring. Given that this PR disentangles  `eth-simple-keyring` from `eth-hd-keyring` and the latter no longer inherits methods from the former (`getAppKeyAddress` , `exportAccount`, `signTransaction`, `signMessage`, `signPersonalMessage`, `decryptMessage`, `signTypedData`, `removeAccount`, `getEncryptionPublicKey`) the methods are now implemented on `eth-hd-keyring` now too.

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -104,26 +121,149 @@ class HdKeyring extends SimpleKeyring {
...
+    if (opts.withAppKeyOrigin) {
+      const { privateKey } = wallet;
+      const appKeyOriginBuffer = Buffer.from(opts.withAppKeyOrigin, 'utf8');
+      const appKeyBuffer = Buffer.concat([privateKey, appKeyOriginBuffer]);
+      const appKeyPrivateKey = arrToBufArr(keccak256(appKeyBuffer, 256));
+      const appKeyPublicKey = privateToPublic(appKeyPrivateKey);
+      wallet = { privateKey: appKeyPrivateKey, publicKey: appKeyPublicKey };
+    }
+
+    return wallet;
```
Comment (adonesky1) [inline index.js:266]: These methods are all copied over from `eth-simple-keyring` since `eth-hd-keyring` no longers inherits them - `eth-simple-keyring` is no longer a parent class to `eth-hd-keyring`. At some point [we will have a `base-keyring` class](https://github.com/MetaMask/types/pull/5) which may implement these methods (or it may be an abstract class which `eth-hd-keyring` will inherit from) as a parent again. But for now they need to live directly on this class.

**Review by mcmire — COMMENTED** [review]
Hey @adonesky1 I know this is in draft but I was curious about a couple of things:

- I can understand the motivation for removing the dependency on `eth-simple-keyring`, but what do we feel about defining some kind of interface that all Keyring classes must follow?
- Do we plan on converting this repo to TypeScript at any point? I guess that could solve the previous point.

**Comment by adonesky1** [comment]
> I can understand the motivation for removing the dependency on eth-simple-keyring, but what do we feel about defining some kind of interface that all Keyring classes must follow?

yes for sure! Per my comment [here](https://github.com/MetaMask/eth-hd-keyring/pull/70#discussion_r1030983856), we will be creating an abstract class/type that will define base/required method signatures for all keyring types @mcmire

**Comment by mcmire** [comment]
@adonesky1 Ah sorry missed that. Makes sense!

**Comment by adonesky1** [comment]
@Gudahtt no I think you're right. I let scope creep a ton here. Will break up into 2 or 3 PRs

File: yarn.lock [hunk 2]
```diff
@@ -312,19 +312,386 @@
...
+  version "5.7.0"
+  resolved "https://registry.yarnpkg.com/@ethersproject/wordlists/-/wordlists-5.7.0.tgz#8fb2c07185d68c3e09eb3bfd6e779ba2774627f5"
+  integrity sha512-S2TFNJNfHWVHNE6cNDjbVlZ6MgE17MIxMbMg2zv3wn+3XSJGosL1m9ZVv3GXCf/2ymSsQ+hRI5IzoMJTG6aoVA==
+  dependencies:
+    "@ethersproject/bytes" "^5.7.0"
+    "@ethersproject/hash" "^5.7.0"
+    "@ethersproject/logger" "^5.7.0"
+    "@ethersproject/properties" "^5.7.0"
+    "@ethersproject/strings" "^5.7.0"
+
```
Comment (adonesky1) [inline yarn.lock:694]: just a note, these are all devDependencies brought in via `@ethereumjs/tx` so they aren't harming bundle reduction efforts

**Comment by adonesky1** [comment]
@Gudahtt this is now just bumping `eth-sig-util` version and removing inheritance of `eth-simple-keyring`. The removal of `ethereumjs-wallet` will be in a separate PR.

**Comment by adonesky1** [comment]
@Gudahtt and @mcmire and it's ready for review

File: package.json [hunk 3]
```diff
@@ -26,13 +26,14 @@
     "test": "jest"
   },
   "dependencies": {
-    "@metamask/eth-sig-util": "^4.0.0",
+    "@ethereumjs/util": "^8.0.2",
```
Comment (Gudahtt) [inline package.json:29]: It looks like this PR includes updating from v7 to v8. Making a note to myself here to review the API changes.

**Review by Gudahtt — COMMENTED** [review]
Thanks for taking care of this! This should make the TypeScript conversion easier.

It was a bit hard to follow how this compared to `eth-simple-keyring@4.2.0` because it seems like some changes were brought from the latest version of that package, and some were from v4.2.0. For example, all of the async functions are now using `async` instead of returning things wrapped in `Promise.resolve`.

So far I've made a lot of notes of changes to document in the changelog. There are a few things I still need to review more closely, I'll try to finish this review soon.

File: index.js [hunk 4]
```diff
@@ -120,13 +135,136 @@ class HdKeyring extends SimpleKeyring {
...
+
+  _getPrivateKeyFor(address, opts = {}) {
+    if (!address) {
+      throw new Error('Must specify address.');
+    }
+    const wallet = this._getWalletForAccount(address, opts);
+    return wallet.privateKey;
+  }
+
+  _getWalletForAccount(account, opts = {}) {
```
[the remark this answers is not in the document]
Comment (adonesky1) [reply index.js:244]: > It looks like this change was what brought in the ethereum-cryptography dependency as well. Was this leftover from the ethereumjs-wallet replacement work?

yes good catch

[the remark this answers is not in the document]
Comment (adonesky1) [reply index.js:244]: Well actually we want to use `ethereum-cryptography`'s version of `keccak256` rather than `ethereumjs-util`'s. This change was made in `eth-simple-keyring` [here](https://github.com/MetaMask/eth-simple-keyring/commit/78bade85f93a42a330171dea31d371a5484abfbe#diff-e727e4bdf3657fd1d798edcd6b099d6e092f8573cba266154583a746bba0f346R196)

File: package.json [hunk 5]
```diff
@@ -26,13 +26,14 @@
...
+    "@metamask/eth-sig-util": "^5.0.2",
     "@metamask/scure-bip39": "^2.0.3",
-    "eth-simple-keyring": "^4.2.0",
-    "ethereumjs-wallet": "^1.0.1"
+    "ethereum-cryptography": "^1.1.2",
+    "ethereumjs-wallet": "^1.0.2"
```
[the remark this answers is not in the document]
Comment (adonesky1) [reply package.json:33]: yeah this change isn't required. Will push a revert.

File: package.json (hunk 3 above)
Comment (adonesky1) [reply package.json:29]: do you mean from v4 to v8?

**Comment by adonesky1** [comment]
> It was a bit hard to follow how this compared to eth-simple-keyring@4.2.0 because it seems like some changes were brought from the latest version of that package, and some were from v4.2.0. For example, all of the async functions are now using async instead of returning things wrapped in Promise.resolve.

@Gudahtt Could you point out which bits you are seeing that are `eth-simple-keyring` v4.2.0 rather than latest. I was shooting to bring over (adapted) copies from only the latest version of `eth-simple-keyring` (`@metamask/eth-simple-keyring` v5.0.0)

File: package.json (hunk 5 above)
[the remark this answers is not in the document]
Comment (adonesky1) [reply package.json]: done [here](https://github.com/MetaMask/eth-hd-keyring/pull/70/commits/e6a04ab40a3622f01cf26f41a7fe068ae44419f2)

**Comment by Gudahtt** [comment]
Oh, maybe it is all from v5. I thought it was a mix because I was comparing it to v4.2.0 and noticed the order the functions matched v4.2.0 more closely than the main branch, I didn't look at v5 directly.

File: package.json (hunk 3 above)
Comment (Gudahtt) [reply package.json:29]: Nope, I mean `@ethereumjs/util@8.0.2`. It was `ethereumjs-util@7` in `eth-simple-keyring`

Comment (adonesky1) [reply package.json:29]: oh whoops yes you're totally right

Comment (adonesky1) [reply package.json:29]: I can also separate out this change if it will make reviewing easier

Comment (Gudahtt) [reply package.json:29]: Hmmm.... would it be reasonably easy to update this package to v5 of `eth-simple-keyring` first then drop the dependence on it? That would bring in most of these changes all in one go.

Comment (adonesky1) [reply package.json:29]: Yeah for sure! No problem

Comment (adonesky1) [reply package.json:29]: done here: https://github.com/MetaMask/eth-hd-keyring/pull/71

Comment (adonesky1) [reply package.json:29]: @Gudahtt #71 has been merged and this PR has been rebased on top of it now
